app/Wrappers/SecretStreamWrapper.php

```diff
@@ -0,0 +1,187 @@
+     * @param $options some options encoded as bits
+     * @param $open
+     * @return bool whether opening was successful
+     */
+    public function stream_open($path, $mode, $options, &$open) {
+        // Check that we are in write mode.
+        if ($mode !== "w") {
+            if ($options & STREAM_REPORT_ERRORS) {
+                trigger_error('Secret stream wrapper may only be opened in "w" mode');
+            }
+            return false;
+        }
+
+        // Open wrapped resource. TODO : It would be nice if our parsing was more robust.
+        $parts = explode("://", $path);
+        if (count($parts) < 2) {
+            if ($options & STREAM_REPORT_ERRORS) {
+                trigger_error('Unable to parse wrapper-received URL');
+            }
+            return false;
+        }
+        $wrappedPath = implode("://", array_slice($parts, 1));
+        if (!($this->wrapped = fopen($wrappedPath, $mode))) {
+            if ($options & STREAM_REPORT_ERRORS) {
+                trigger_error('Wrapped path could not be opened');
+            }
+            return false;
+        }
+
+        // Initialise secret stream.
+        list($this->stream, $header) = sodium_crypto_secretstream_xchacha20poly1305_init_push(self::getKey());
+
+        // Write out initial header.
+        if (!fwrite($this->wrapped, $header)) {
+            if ($options & STREAM_REPORT_ERRORS) {
+                trigger_error('IO error when writing header');
+            }
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Handles writes to the resource, which we will encrypt. e.g. `fwrite()` would call this.
+     *
+     * @param $data the plaintext to write
+     * @return int how many bytes of input we used, else 0 if everything went wrong
+     */
+    public function stream_write($data) {
+        // Append data to our internal buffer.
+        $this->buffer .=  $data;
+
+        // Write out as many full-sized messages as we can without fully depleting the buffer.
+        if ($this->flushBuffer()) {
+            // We wrote everything provided to us.
+            return strlen($data);
+        }
+
+        // We failed to write.
+        return 0;
+    }
+
+    /**
+     * Flush out our final data and close all resources. e.g. `fclose()` would call this.
+     */
+    public function stream_close() {
+        // Write out data as encrypted messages until we have only the final message's data in the buffer.
+        $this->flushBuffer();
+
+        // Flush out the final message, tagged as such.
+        if (strlen($this->buffer) > 0) {
+            $encrypted = sodium_crypto_secretstream_xchacha20poly1305_push(
+                $this->stream,
+                $this->buffer,
+                '', // This is an inactive placeholder.
+                SODIUM_CRYPTO_SECRETSTREAM_XCHACHA20POLY1305_TAG_FINAL
+            );
+            fwrite($this->wrapped, $encrypted); // TODO : Can an error here be caught somehow?
```

Comment:
frwite, being a fairly basic function returns falsey (false or, possibly 0) as an error state. not terribly useful

Reply:
From testing, after writing the comment, it seemed that Laravel was eager to treat these errors pretty seriously without having to spot them manually. The error would be spotted by the single user on serving the file as well, so I think I'll leave this uncaught :+1: